### PR TITLE
Swap integration tests round so each file is runnable on its own

### PR DESCRIPTION
**Pull request by DanTup**
This makes it possible to `pub run test test/integration/logging_test.dart` for example, and also works with the VS Code test runner tree/CodeLens links.

I committed the renames as one change then the changes in the hope that GH will show the diff better than a load of deletes/new files...

We might not care about run_all.dart, I just kept it (after renaming) because previously it was possible to run all integration tests from a script like that. I'll delete it if we think it's extra maintenance.

### CODE REVIEW

**Comment by DanTup** [comment]
Oops, apparently I sorted one of the files making the diff bigger; lemme fix...

**Comment by DanTup** [comment]
Ok, GH made a hash of the combined diff so probably the best way to review is:

1. Look at the first changeset which is just renames with no changes
2. Review the second changeset directly with whitespace changes ignored and side-by-side
  https://github.com/flutter/devtools/pull/232/commits/bb74bddfd6aac87bc28f5b557f0b9c511b2ef56d?utf8=%E2%9C%93&diff=split&w=1

Each test file basically has its own `main()` method now, `run_all.dart` is a script for running all (as was `integration_test.dart` before) and the shared setup is still shared but needs to be called explicitly.

File: packages/devtools/test/integration_tests/app_test.dart [hunk 1]
```diff
@@ -7,47 +7,52 @@ import 'package:test/test.dart';
...
-    // make sure the connect dialog displays
-    await waitFor(() async => await connectDialog.isVisible());
-
-    // have it connect to a port
-    await connectDialog.connectTo(appFixture.servicePort);
-
-    // make sure the connect dialog becomes hidden
-    await waitFor(() async => !(await connectDialog.isVisible()));
+void main() {
+  group('app', () {
```
Comment (jacob314) [inline packages/devtools/test/integration_tests/app_test.dart]: what is the benefit of adding this group?

File: packages/devtools/test/integration_tests/debugger_test.dart [hunk 2]
```diff
@@ -155,132 +145,151 @@ void debuggingTests() {
...
-  });
+      // verify state resumed
+      expect(await debuggingManager.getState(), 'running');
+    });
+
+    test('console output', () async {
+      appFixture = await CliAppFixture.create(
+          'test/fixtures/debugging_app_exception.dart');
+
+      final DevtoolsManager tools =
```
Comment (jacob314) [inline packages/devtools/test/integration_tests/debugger_test.dart]: this diff is quite hard to read. Are there any real changes here?

File: packages/devtools/test/integration_tests/app_test.dart (hunk 1 above)
Comment (DanTup) [reply packages/devtools/test/integration_tests/app_test.dart]: I figured it'd keep the test names as before with a prefix - though it just occurred to me that you see the filename too, so probably it's superfluous now - I can remove :-)

File: packages/devtools/test/integration_tests/debugger_test.dart (hunk 2 above)
Comment (DanTup) [reply packages/devtools/test/integration_tests/debugger_test.dart]: Are you looking at the PR, or the diff from the link in my comment above? Hopefully the link in the comment will give you a clean diff :-)

File: packages/devtools/test/integration_tests/run_all.dart [hunk 3]
```diff
@@ -3,33 +3,14 @@
...
-    tearDownAll(() async {
-      await browserManager?.teardown();
-      await webdevFixture?.teardown();
-    });
-
-    group('app', appTests);
-    group('logging', loggingTests);
-    group('debugging', debuggingTests);
-  }, timeout: const Timeout.factor(2));
+  app.main();
```
Comment (jacob314) [inline packages/devtools/test/integration_tests/run_all.dart]: I think it would be cleaner to leave this file as is but add main methods to the individual files that allow executing just those tests as needed.
I suspect this may lead to surprising behavior if the setUpAll and tearDownAll calls for the different tests ever diverged.

Comment (DanTup) [reply packages/devtools/test/integration_tests/run_all.dart]: Do you mean put the tests in the other files back into their own functions, then call those functions both from here and from the main's inside there?

Comment (jacob314) [reply packages/devtools/test/integration_tests/run_all.dart]: Ok the ignore whitespace diff is great! I'd just been manually looking at the commits after the commit to move the files.

**Comment by DanTup** [comment]
So, I removed the useless groups and I deleted `run_all.dart` on the basis that without the setup/teardown in it there was no value (and you can run all integration tests with `pub run test test/integration`).

However, it just occurred to me that they're run in different isolates now so the setup is all run for each test. This is what @devoncarew meant when he said about sharing the fixtures - I totally overlooked that they were in their own isolates (assuming they could share top-level variables).

So, now I'm less sure about this. Running individual files/tests works much better with this (eg. using VS Code's test runner), however running *all of the tests* is much slower.

We can put back the `run_all.dart` and have "extra" main methods in each test file, but I think that might be a bit of a maintenance pain. Any thoughts?

**Comment by devoncarew** [comment]
I think the code was a little simpler before, and the test cycle times faster. Extra `main` methods would work, but I don't know how often people will be running the integration tests suites individually.

**Comment by DanTup** [comment]
For running all tests it was definitely faster, but was more complicated to run an individual test (and the VS Code test runner + CodeLens links are broken). I had another possible idea to fix this, but I don't know how reliable it'd be here:

https://github.com/Dart-Code/Dart-Code/issues/1439#issuecomment-459311259

The idea is if a test file doesn't end with `_test.dart` then run it without the filename. That means if the name overlaps with tests in other files, they'll run too, but might at least restore this if they're unique.

WDYT? Or can you think of another way?

**Comment by devoncarew** [comment]
Hey - I don't have a strong opinion here :) I think the two things I care about are keeping our tests reasonably simple (there's already some complexity in getting chrome, webdev serve, and the fixture app all set up and talking), and that are bot turnaround times on travis are reasonably speedy. If you have a refactor to the tests that you'd like to land, that don't hugely impact test simplicity or bot times, lgtm.

**Comment by DanTup** [comment]
I think we have two goals that don't currently work together with the `test` package:

1. Ability to have tests from multiple files run in a single isolate (to allow sharing setup/teardown without putting all the tests in one file)
2. Ability to run an individual test from an IDE

To achieve 1, we need to define tests in a non-`_test.dart` file and include them from a `_test.dart` file. However, an IDE doesn't have the required information when looking at `foo.dart` to be able to run an individual test for it (it would need to know to run it from `some_test.dart` and how to construct a `--name` that would run only that test).

I think the closest I can get is that in Dart Code if you open a file that does not end with `_test.dart` but is in the `test` folder, I run it with just `pub run test` with no filename, and do a partial match on the name (normally we anchor to start/end to reduce overlaps), however that would likely result in running more tests than the user wanted if they have a lot of overlaps in names.

@jjakemac53 @grouma Do you know if there's a way to achieve what we want?

If there was something like `pub run test --only-from-file foo.dart` which would do the same as `pub run test` (eg. load all test files) but only execute any tests that were defined in `foo.dart` (even though there was no `main` for `foo.dart`) that would work?

**Comment by jakemac53** [comment]
The new structure is definitely the most idiomatic package:test structure. I don't think we want to add the extra logic to try and handle this use case in package:test, it would be quite difficult to support well I think.

Note that a lot of the slowdown could just be the extra kernel compilation, not the setup/teardown calls. If it is mostly kernel compilation (this shows as the "loading" time on the console) then you can try running tests with build_runner - but depending on the patterns used in the test it might not always work out of the box. You would need to add the build_runner, build_vm_compilers, and build_test packages as dependencies, and then run `pub run build_runner test`. You also would want to cache the `.dart_tool` directory on travis to get the benefit there.

**Comment by DanTup** [comment]
> Note that a lot of the slowdown could just be the extra kernel compilation

In these integration tests we're spawning a server (`pub global run webdev serve web`) and firing up Chrome w/ dev tools connection in `setupAll` which is pretty slow. Having the tests split across 4 files that all do that adds a lot of time on (I haven't timed it, but I suspect these dwarf the compilation of the Dart test code).

**Comment by jakemac53** [comment]
> In these integration tests we're spawning a server (`pub global run webdev serve web`) and firing up Chrome w/ dev tools connection in `setupAll` which is pretty slow. Having the tests split across 4 files that all do that adds a lot of time on (I haven't timed it, but I suspect these dwarf the compilation of the Dart test code).

Ah ok, then I would probably just stick with essentially what you have now.

One thing you could do, although a bit ugly, is restructure the code so there is an actual `test` call in the root file for each test, and you just pass a reference to the actual method performing the test to each of those. That way the tests are defined in the root file and I think existing logic in ides would be able to pick it up. The downside is you have to add some boilerplate in the root test file for every test, so adding tests requires editing two files.

**Comment by DanTup** [comment]
> Ah ok, then I would probably just stick with essentially what you have now.

The problem with what we have now is that IDE test runners can't run the tests, because they don't know that to run tests in `loggging.dart` they need to invoke `pub run test integration_test.dart` :(

Would you consider a feature that allows passing a filename to `pub run test` but that's used only to filter tests, and not used as the entry point? The CodeLens links for running tests in VS Code are really useful, but for some projects (like this) you wait several seconds just to get an error spat out to the console.

**Comment by jakemac53** [comment]
> The problem with what we have now is that IDE test runners can't run the tests, because they don't know that to run tests in `logging.dart` they need to invoke `pub run test integration_test.dart` (and even if they did, it's hard to construct a filter to just get that one test) :(

Right, what I am saying is this is an explicit tradeoff you can make for this package to make the tests run faster. You sacrifice the ability to run tests via the ide (for this repo only), to get the benefit of running the tests faster.

I did just think of a different pattern that could maybe work though, what if you used part files instead? That way the IDE can know the parent test file for a given test, via the `part of` directive.

**Comment by jakemac53** [comment]
> This isn't the first time this has come up, but normally I suggest people split their tests up into separate files and they've been happy to.

This is really specific to large integration tests though, where you want to share some work between them but also keep the actual test implementations in separate files. I think that is different from the typical problem where in Dart 1 the only reasonable way to run tests was to create an all_tests.dart file that manually imports and runs all your separate test files. In that case just separating the tests out works generally fine.

If you look at package:test in isolation, this isn't really a problem. The existing pattern in this repo works just fine and you can run tests with their name explicitly to only run certain tests (maybe not as easy as running the test file, but a reasonable tradeoff).

The problem I think is only how that interacts with an IDE that wants to provide debug/run hooks on individual tests within a file when you are looking at it. In most cases, all the calls to `test` exist in the original test entrypoint, but there is a general problem any time that is not the case. The run/debug links are going to exist in that shared library file, which isn't itself runnable. I don't really see how the IDE can easily resolve this scenario, and I don't know that its actually worth even trying to resolve it to be completely honest.

The reasons I don't think its worth resolving are:

- It is a rare edge case
  - It only blocks one specific feature in ide from working, but doesn't "break" anything
  - The scenario to reproduce it is also rare in itself
- There are workarounds available
  - Move all the tests to a single file
  - Separate the tests and accept the slower run times

**Comment by DanTup** [comment]
> I think that is different from the typical problem where in Dart 1 the only reasonable way to run tests was to create an all_tests.dart file that manually imports and runs all your separate test files

Ah, I didn't realise this was a Dart1 problem. I'd seen it done a lot and assumed people just liked doing it this way.

> I don't really see how the IDE can easily resolve this scenario

I don't think the IDE can in isolation, but I think maybe with some supporting changes in `test` it could :-) Is the suggestion above (to support a filter based on the filename) no good?

It's getting a bit off-topic for this PR, but I try *really* hard to make things "just work" in Dart Code. It makes me a little sad when there are buttons people can click and they just error and the answer is to set things up in some specific way (even though the existing way "works fine outside of Dart Code"). The test runner already has a lot of quirks that confuse users. I think the easier we can make it for people to run their tests, the more likely they will write tests.

But alas, I understand adding to (and maintaining) new features is expensive (moreso in things that have lots of downstream dependents) but I thought it was at least worth raising for discussion. I'll see if I can find a better way to handle this in Dart Code to avoid the CodeLens links where they won't work (probably anything that doesn't end `_test.dart`).

Thanks for the discussion anyway!
